drivers/tee/tee_shm.c

```diff
@@ -190,17 +207,109 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
 }
 EXPORT_SYMBOL_GPL(tee_shm_alloc);
 
+struct tee_shm *tee_shm_register_fd(struct tee_context *ctx, int fd)
+{
+	struct dmabuf_ref *ref = NULL;
+	struct tee_shm *shm;
+	void *rc;
+	DEFINE_DMA_BUF_EXPORT_INFO(exp_info);
+
+	if (!tee_device_get(ctx->teedev))
+		return ERR_PTR(-EINVAL);
+
+	shm = kzalloc(sizeof(*shm) + sizeof(*ref), GFP_KERNEL);
+	if (!shm) {
+		rc = ERR_PTR(-ENOMEM);
+		goto err;
+	}
+
+	shm->ctx = ctx;
+	shm->teedev = ctx->teedev;
+	shm->id = -1;
+	ref = (struct dmabuf_ref *)(shm + 1);
+
+	ref->dmabuf = dma_buf_get(fd);
+	if (!ref->dmabuf) {
+		rc = ERR_PTR(-EINVAL);
+		goto err;
+	}
+
+	ref->attach = dma_buf_attach(ref->dmabuf, &shm->teedev->dev);
+	if (IS_ERR_OR_NULL(ref->attach)) {
+		rc = ERR_PTR(-EINVAL);
+		goto err;
+	}
+
+	ref->sgt = dma_buf_map_attachment(ref->attach, DMA_BIDIRECTIONAL);
+	if (IS_ERR_OR_NULL(ref->sgt)) {
+		rc = ERR_PTR(-EINVAL);
+		goto err;
+	}
+
+	if (sg_nents(ref->sgt->sgl) != 1) {
+		rc = ERR_PTR(-EINVAL);
+		goto err;
+	}
+
+	shm->paddr = sg_dma_address(ref->sgt->sgl);
+	shm->size = sg_dma_len(ref->sgt->sgl);
+	shm->flags = TEE_SHM_DMA_BUF;
+
+	mutex_lock(&shm->teedev->mutex);
+	shm->id = idr_alloc(&shm->teedev->idr, shm, 1, 0, GFP_KERNEL);
+	mutex_unlock(&shm->teedev->mutex);
+	if (shm->id < 0) {
+		rc = ERR_PTR(shm->id);
+		goto err;
+	}
+
+	/* export a dmabuf to later get userland ref */
+	exp_info.ops = &tee_shm_dma_buf_ops;
+	exp_info.size = shm->size;
+	exp_info.flags = O_RDWR;
+	exp_info.priv = shm;
+
+	shm->dmabuf = dma_buf_export(&exp_info);
+	if (IS_ERR(shm->dmabuf)) {
+		rc = ERR_PTR(-EINVAL);
+		goto err;
+	}
+
+	mutex_lock(&shm->teedev->mutex);
+	list_add_tail(&shm->link, &ctx->list_shm);
+	mutex_unlock(&shm->teedev->mutex);
+
+	return shm;
+
+err:
+	if (shm && shm->id >= 0) {
+		mutex_lock(&ctx->teedev->mutex);
+		idr_remove(&ctx->teedev->idr, shm->id);
+		mutex_unlock(&ctx->teedev->mutex);
+	}
+	if (ref && ref->sgt)
+		dma_buf_unmap_attachment(ref->attach, ref->sgt,
+					 DMA_BIDIRECTIONAL);
+	if (ref && ref->attach)
+		dma_buf_detach(ref->dmabuf, ref->attach);
+	if (ref->dmabuf)
```

Comment:
`ref &&` is missing, wouldn't it be better to surround these three ifs with a `if (ref) {...`?